staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go

```
@@ -170,70 +171,79 @@ func (c *Controller) sync(name string) error {
-		// spec that already has defaults pruned
-		spec.Definitions = handler.PruneDefaults(spec.Definitions)
-		if err != nil {
-			return nil, false, err
-		}
-		newSpecs[v.Name] = spec
-		if oldSpecs[v.Name] == nil || !reflect.DeepEqual(oldSpecs[v.Name], spec) {
-			anyChanged = true
-		}
+		newSpecs[v.Name] = cached.NewStaticSource(func() cached.Result[*spec.Swagger] {
+			s, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true})
+			// Defaults must be pruned here for CRDs to cleanly merge with the static
+			// spec that already has defaults pruned
+			if err != nil {
+				return cached.NewResultErr[*spec.Swagger](err)
+			}
+			s.Definitions = handler.PruneDefaults(s.Definitions)
+			return cached.NewResultOK(s, fmt.Sprintf("%d", crd.Generation))
+		})
 	}
-	if !anyChanged && len(oldSpecs) == len(newSpecs) {
-		return newSpecs, false, nil
-	}
-
-	return newSpecs, true, nil
+	return newSpecs, nil
 }
 
-// updateSpecLocked aggregates all OpenAPI specs and updates openAPIService.
-// It is not thread-safe. The caller is responsible to hold proper lock (Controller.lock).
-func (c *Controller) updateSpecLocked() error {
-	crdSpecs := []*spec.Swagger{}
-	for _, versionSpecs := range c.crdSpecs {
-		for _, s := range versionSpecs {
-			crdSpecs = append(crdSpecs, s)
+func (c *Controller) buildMergeSpecLocked() error {
+	crdSpecs := map[string]cached.Data[*spec.Swagger]{}
+	for g := range c.crdSpecs {
+		for v := range c.crdSpecs[g] {
+			crdSpecs[g+"/"+v] = c.crdSpecs[g][v]
 		}
 	}
-	mergedSpec, err := builder.MergeSpecs(c.staticSpec, crdSpecs...)
-	if err != nil {
-		return fmt.Errorf("failed to merge specs: %v", err)
-	}
-	return c.openAPIService.UpdateSpec(mergedSpec)
+
+	cache := cached.NewMerger(func(results map[string]cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		localCRDSpec := []*spec.Swagger{}
+		for k := range results {
+			localCRDSpec = append(localCRDSpec, results[k].Data)
```

Not checking for errors? You wouldn't want to put some `nil` in there

---

Added a check for errors. I can't really think of a case where this would fail unless the CRD was malformed. If it's malformed, it shouldn't have passed validation to be applied to the server to begin with. Also a bit unsure of how we should surface the error, an error log would probably be best?